src/core/stdcpp/exception.d

```diff
@@ -87,18 +87,18 @@ else version (CppRuntime_Microsoft)
     {
     @nogc:
         ///
-        this(const(char)* message = "unknown", int = 1) nothrow { msg = message; }
+        extern(D) this(const(char)* message = "unknown", int = 1) nothrow { msg = message; }
         ///
-        extern(D) ~this() nothrow {}
+        @weak ~this() nothrow {}
```

Comment:
I'd like to be clear on the effect of these changes.

I guess that marking it `@weak` will reduce clutter in the binary because rather than having 2 copies with different symbol names, it will throw one out, which is good. I've been begging for `@weak` for 12 years or something!

But that being the case, what's with the constructor above? It looks like I didn't mark it `extern(D)` before, which means it might cause a link error if was also emit in a C++ module (is that why it was changed? I must have thought it was unable to collide for some reason to not have marked it)... but why change that to `extern(D)` instead of also making it `@weak` like these other changes?

Reply:
According to the introducing commit, it was switched to `extern(D)` to fix D/C++ symbol conflicts and undefined D vtable entries.

Given that the constructors are `final` we can own the ABI of it.  This is not the case for the destructors, which are marked `virtual` on the C++ side, so they needed to be switched to `extern(C++)` to prevent mismatching the ABI.

Reply:
Are there cases where ABI mismatch occurred? I didn't realise there was an ABI mismatch possible for virtual methods with signatures like these?

Reply:
The Windows OMF pipelines report signature mismatch errors at link-time.

There's also this issue for some more context https://github.com/ldc-developers/ldc/issues/3530

And this PR fo rsome more https://github.com/dlang/druntime/pull/2788

Reply:
> Nice! With those, will things still be tied to the `CppRuntime_` druntime was compiled with ?

Yes, though IIUC Clang and GCC are compatible?  Probably the only two we care about in terms of movability because Microsoft doesn't use the Itanium ABI.

Reply:
@TurkeyMan I'm also noting that these constructors are defined inline in the C++ headers, so they are not directly callable from D anyway.